penzai/core/selectors.py

```diff
@@ -39,6 +39,44 @@
 T = typing.TypeVar("T")
 
 
+def _shift_negative_indices(indices: Iterable[int], shift: int) -> tuple[int, ...]:
+  """Adds `shift` to negative indices and leaves non-negative indices unchanged
+
+  Can be used to handle negative indices. For example, if we expect indices in
+  `r = range(6)` and we get `[0, 3, -2]` as input, we can use
+
+  ```py
+  shift_negative_indices([0, 3, -2], len(r))
+  ```
+
+  to get `(0, 3, 4)`. The same can be achieved in more generality with
+
+  ```py
+  pz.select((0, 3, -2)) \
+    .at_instances_of(int) \
+    .where(lambda i: i < 0) \
+    .apply(lambda i: i + shift)
```

Comment:
Not sure if I should just use this instead of defining `_shift_negative_indices`. I'm weary of doing so because we're in the module that defines selectors here. Also, we don't need that kind of generality at this point and `_shift_negative_indices` avoids some overhead

Reply:
Defining a new function here like you've done makes sense to me, the logic is probably too simple to benefit much from using selectors. (You could also just inline the logic into the call site since it's only used once.)

Reply:
> You could also just inline the logic into the call site since it's only used once

I'd prefer to keep it separate so code blocks stay short and hence easy to interpret